[PM-10749] [BEEEP] New export format: Zip with attachments

quexten opened the pull request:
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-10749

https://community.bitwarden.com/t/allow-attachments-to-be-exported-when-using-export-data/835/1

## 📔 Objective

This PR adds a new export format - zip - which includes all attachments from the personal vault.
The layout is as follows:

- data.json
- attachments/{cipherid}/attachment1.pdf
- attachments/{cipherid}/attachment2.jpg
- attachments/{cipherid2}/attachmentabc.txt

The zip file is assembled on the client side in order to have a decrypted, fully usable export without external tools converting (decrypting) them.

#### Of note:
- Org exports are not supported yet because they could be too large
- We currently create the zip file in memory. This is fine for personal vaults, but we need a different library for streaming this data to disk (e.g. https://gildas-lormeau.github.io/zip.js/)
- The zip library we use does not support encryption (also https://gildas-lormeau.github.io/zip.js/ would help here)
- No importing is added. Since the data.json is included, that can be imported for now, but we could add full zip + attachment import in a second step

## 📸 Screenshots
[Screencast from 2024-08-09 13-14-06.webm](https://github.com/user-attachments/assets/1f63e2ae-5a5f-4f2e-a3e3-5af7cec657dd)



## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

quexten commented:
Also, do we know how much users / orgs actually store in attachments? I would suspect for large file storage they would use a different service, so we might just enable org attachment export too.

quexten committed:
Add new export format: zip

djsmith85 commented:
> @djsmith85 Since zips are intransparent blobs (unlike json), is there any good way to add tests for this? Maybe unzip the created blob again using the same library, and compare the files?

Yes, that would be a good approach. I wouldn't go as far comparing the files (content) as just checking that it contains a data.json and the expected filenames for attachments.

If we are sticking with JSZip for now that would just be JSZip.loadAsync() and then iterating/searching for the files. 

One of the reasons I chose JSZip at the time was support for JS and Node execution (cli), but I'm totally fine with using another library, especially if it's offering better support for streams.

quexten committed:
Restrict zip export to just individual vaults

quexten committed:
Add tests

quexten committed:
Merge branch 'main' into auth/beeep/zip-export

quexten commented:
Added tests for data.json being present, and for one attachment being present. Sadly I had to mock request, fetch on the global object. I think in the mid-term, creating something like an "AttachmentDownloadService" might make sense since we have fetch calls just spread throughout the code atm.

quexten committed:
Remove unused import

audreyality commented:
> @djsmith85 Since zips are intransparent blobs (unlike json), is there any good way to add tests for this? Maybe unzip the created blob again using the same library, and compare the files?

You might also check how the library performs it's tests.
https://github.com/Stuk/jszip/blob/main/test/asserts/file.js

Teasing apart what those helpers do might be a bit much, though.

quexten committed:
Merge branch 'main' into auth/beeep/zip-export

quexten committed:
Merge branch 'main' into auth/beeep/zip-export

quexten committed:
Fix build error

quexten committed:
Fix tests

quexten committed:
Merge branch 'main' into auth/beeep/zip-export

quexten committed:
Fix test

quexten committed:
Merge branch 'main' into auth/beeep/zip-export

quexten committed:
Merge branch 'main' into auth/beeep/zip-export

croci commented:
Please merge this! The zip export would be extremely useful!

quexten committed:
Merge branch 'main' into auth/beeep/zip-export

djsmith85 committed:
Merge branch 'main' into auth/beeep/zip-export

djsmith85 committed:
Fix retrieval of ciphers by passing in activeUserId

djsmith85 committed:
Merge branch 'main' of https://github.com/bitwarden/clients into auth/beeep/zip-export

djsmith85 committed:
Merge branch 'main' of https://github.com/bitwarden/clients into auth/beeep/zip-export

audreyality commented on apps/cli/src/tools/export.command.ts:
```diff
@@ -62,7 +62,7 @@ export class ExportCommand {
       exportContent =
         options.organizationid == null
-          ? await this.exportService.getExport(format, password)
+          ? ((await this.exportService.getExport(format, password)) as string)
```
⚠️ This needs to test whether the output is actually a string. Alternatively, consider splitting `getExport` into a unique method per return type.

audreyality commented on libs/tools/export/vault-export/vault-export-core/src/services/individual-vault-export.service.ts:
```diff
@@ -46,18 +48,53 @@ export class IndividualVaultExportService
   }
 
-  async getExport(format: ExportFormat = "csv"): Promise<string> {
+  async getExport(format: ExportFormat = "csv"): Promise<string | Blob> {
     if (format === "encrypted_json") {
       return this.getEncryptedExport();
+    } else if (format === "zip") {
+      return this.getDecryptedExportZip();
     }
```
🤔 Seems like the `Blob` return was introduced by `this.getDecryptedExportZip()`. Since the other two methods return strings, it seems reasonable to expect `getDecryptedExportZip` to be called from its own code-path.

💭 Alternatively, `getEncryptedExport` and `getDecryptedExport()` could have their result be converted to returning `Blob`, which would eliminate the string-typed return from `getExport()`.

djsmith85 replied:
I don't disagree, although I think will be kept as is for now. https://github.com/bitwarden/clients/pull/13842 definitely improves the weirdness between string | Blob returns.

The way it is currently implemented is by introducing a new format (`zip`). In the future I imagine supporting attachments for any format (`csv`, `json`, `encrypted_json`) and including them would be an option to the user (`includeAttachments: boolean`).

audreyality commented on libs/tools/export/vault-export/vault-export-ui/src/components/export.component.ts:
```diff
@@ -452,12 +459,21 @@ export class ExportComponent implements OnInit, OnDestroy, AfterViewInit {
+
+  private downloadZipFile(blob: Blob): void {
+    const fileName = this.getFileName();
+    this.fileDownloadService.download({
+      fileName: fileName,
+      blobData: blob,
+      blobOptions: { type: "application/zip" },
+    });
+  }
```
🎨 Since the data's content type is required for proper interpretation of the result, consider returning it with the exported data.

```typescript
// in vault-export.service.abstraction, introduce abstraction
type ExportedVault = {
  type: "application/zip",
  data: Blob,
  name: string // should the file name be attached here as well?
} | {
  type: "text/plain",
  data: string,
  name: string // should the file name be attached here as well?
};

// update `getExport` prototype w/ abstraction
getExport: (format: ExportFormat, password: string) => Promise<ExportedVault>;
```

In the above code, `ExportedVault.type` is a property enabling type selectivity on `ExportedVault.data` using [discriminated union](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions) logic.

djsmith85 replied:
Awesome idea and I created https://github.com/bitwarden/clients/pull/13842 which is stacked on top off this PR, to make it easier to review.

audreyality commented:
> Yes, that would be a good approach. I wouldn't go as far comparing the files (content) as just checking that it contains a data.json and the expected filenames for attachments.

🎨 You could, however, check to see that the attachment files are listed in the `data.json`.

djsmith85 committed:
Merge branch 'main' of https://github.com/bitwarden/clients into auth/beeep/zip-export

djsmith85 commented on libs/tools/export/vault-export/vault-export-core/src/services/individual-vault-export.service.ts:
```diff
@@ -46,18 +48,53 @@ export class IndividualVaultExportService
+    const activeUserId = await firstValueFrom(this.accountService.activeAccount$.pipe(getUserId));
+
+    // attachments
+    for (const cipher of await this.cipherService.getAllDecrypted(activeUserId)) {
```
⚠️ The list of ciphers retrieved here is not properly filtered. It doesn't skip ciphers with an organizationId assigned to it. Adding that to the if below, would solve this temporarily.

🌱 A larger change, but a better way would be: Consider breaking up the retrieval and filtering done within getDecryptedExport() and only then passing in the list of

djsmith85 commented on libs/tools/export/vault-export/vault-export-core/src/services/individual-vault-export.service.ts:
```diff
@@ -46,18 +48,53 @@ export class IndividualVaultExportService
+
+      const cipherFolder = attachmentsFolder.folder(cipher.id);
+      for (const attachment of cipher.attachments) {
+        const response = await fetch(new Request(attachment.url, { cache: "no-store" }));
```
🤔 Something I need to clarify further internally, but the url used here within a development environment, does not point to the place to download an attachment. This might. be a bug within our dev setup and it has gone unnoticed as all other call sites use the CipherController (`apiService.getAttachmentData()`) which returns an `AttachmentResponse` with the correct download location within the url-property.

djsmith85 reviewed (commented):
Huge thanks to @quexten for working on this. I apologize for the long review time, as this has been stuck between other priorities. We've discussed this internally a couple of times and decided that @bitwarden/team-tools-dev will take ownership and carry this PR. Thanks @audreyality for doing an initial review.

I'm adding some notes and code comments on the PR for transparency, while working through the review/development and testing. I'll be working on resolving them here.

Some other things to note:
- ~~There is currently no error-handling (with the download part)~~ 41a1fe7
- There is no visual indication/error popup to the user in case something fails (only console logs)
- There is no progress indication other than the loading spinner after clicking "Export vault". Not an issue prior as the export files tend to be small, now they will be larger and thus taking longer
- The export once triggered cannot be cancelled

Some of the things mentioned above, are true even now without these changes, but with longer running processes and more potential failure points, we should aim to include them, to provide better feedback to the user.

djsmith85 committed:
Merge branch 'main' of https://github.com/bitwarden/clients into auth/beeep/zip-export

djsmith85 committed:
Guard feature behind `export-attachments`-feature-flag

djsmith85 committed:
Extend cipher filter to also filter out any ciphers that are assigned to an organization

djsmith85 committed:
Added apiService to retrieve AttachmentData (metaData) and then download the attachment

- Added ApiService as a depdency within DI for VaultExportService/IndividualVaultExportService
- Added unit tests for filtering ciphers
- Added unit test for downloading attachment metadata and attachments

djsmith85 committed:
Moved attachment decryption into a separate method and added unit tests

djsmith85 committed:
Added null check for creating the base attachment folder

djsmith85 committed:
Move format check for zip within Org export into an early return/throw

djsmith85 committed:
Add feature flag guard on the CLI

djsmith85 committed:
Merge branch 'main' of https://github.com/bitwarden/clients into auth/beeep/zip-export

djsmith85 committed:
Extend ExportScopeCallout to display an individual export will contain attachment when zip-format is selected

djsmith85 committed:
Fix adding/removing the zip-export option based on selected vault and state of `export-attachments` feature-flag

audreyality commented on libs/tools/export/vault-export/vault-export-core/src/services/individual-vault-export.service.spec.ts:
```diff
@@ -242,6 +261,75 @@ describe("VaultExportService", () => {
+      ) as any;
+      global.Request = jest.fn(() => {}) as any;
+
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      expect(async () => {
+        await exportService.getExport("zip");
+      }).rejects.toThrow("Error downloading attachment");
```
```suggestion
      await expect(async () => {
        await exportService.getExport("zip");
      }).rejects.toThrow("Error downloading attachment");
```

djsmith85 replied:
Now part of e5659743154047aa5d6e95e311f44a4b4e823d84

audreyality commented on libs/tools/export/vault-export/vault-export-core/src/services/individual-vault-export.service.spec.ts:
```diff
@@ -242,6 +261,75 @@ describe("VaultExportService", () => {
+      ) as any;
+      global.Request = jest.fn(() => {}) as any;
+
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      expect(async () => {
+        await exportService.getExport("zip");
+      }).rejects.toThrow("Error decrypting attachment");
```
```suggestion
      await expect(async () => {
        await exportService.getExport("zip");
      }).rejects.toThrow("Error decrypting attachment");
```

djsmith85 replied:
Addressed with 4b4c7779c66fdcf329e4d2c8271c059f8a5660dd

audreyality commented on libs/tools/export/vault-export/vault-export-core/src/services/individual-vault-export.service.spec.ts:
🎨 Consider using whitespace to separate your tests according to arrange/act/assert. If you need more than these 3 groups (for example, you need to assert a precondition), then introduce it with a comment.

djsmith85 replied:
Addressed with 3ac6dd0b2fad93507f6abe70f54ff1adbb262613

audreyality commented on libs/tools/export/vault-export/vault-export-core/src/services/individual-vault-export.service.spec.ts:
```diff
@@ -242,6 +261,75 @@ describe("VaultExportService", () => {
+
+      await exportService.getExport("zip");
+
+      expect(cipherService.getAllDecrypted).toHaveBeenCalled();
+      expect(apiService.getAttachmentData).not.toHaveBeenCalled();
```
⚠️ This does not assert that the ciphers have been filtered out. It asserts that specific methods shouldn't be called. Either assert the result of the export or update the test case label.

audreyality commented on libs/tools/export/vault-export/vault-export-core/src/services/individual-vault-export.service.spec.ts:
```diff
@@ -242,6 +261,75 @@ describe("VaultExportService", () => {
+      expect(apiService.getAttachmentData).not.toHaveBeenCalled();
+    });
+
+    it("throws error if attachment can't be downloaded (status !== 200)", async () => {
```
🎨 Consider testing several status values and asserting this test case affirmatively.

```
it.each([[400], [401], [404], [500]])(
  "throws error if the http request fails (status === %n)",
  async (status) => { /* test case */ }
);
```

djsmith85 replied:
Addressed with e5659743154047aa5d6e95e311f44a4b4e823d84

audreyality commented on libs/tools/export/vault-export/vault-export-core/src/services/individual-vault-export.service.ts:
```diff
@@ -70,31 +73,66 @@ export class IndividualVaultExportService
+          ? attachment.key
+          : await this.keyService.getOrgKey(cipher.organizationId);
+      return await this.encryptService.decryptToBytes(encBuf, key);
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    } catch (e) {
+      throw new Error("Error decrypting attachment");
+    }
```
⚠️ You should either use `e` or delete it. You should also assert its type as `unknown`.

🎨 Consider logging `e` using the `SemanticLogger` or `LogService`.

djsmith85 replied:
Addressed with 1c9dd461438d96256e19fc5a9b9e7f89d6465a7c

audreyality commented on libs/tools/export/vault-export/vault-export-ui/src/components/export.component.ts:
```diff
@@ -271,10 +276,20 @@ export class ExportComponent implements OnInit, OnDestroy, AfterViewInit {
-        this.organizationId = value != "myVault" ? value : undefined;
+      .subscribe(([value, isExportAttachmentsEnabled]) => {
+        this.organizationId = value !== "myVault" ? value : undefined;
+        if (value === "myVault" && isExportAttachmentsEnabled) {
+          if (!this.formatOptions.some((option) => option.value === "zip")) {
+            this.formatOptions.push({ name: ".zip (with attachments)", value: "zip" });
+          }
+        } else {
+          this.formatOptions = this.formatOptions.filter((option) => option.value !== "zip");
+        }
```
⛏️ Consider restructuring these `if` statements to reduce nesting.

audreyality reviewed (commented):
There's a few changes you'll need to make to your test cases. Also a few places where you should be following a lint instead of disabling it.

djsmith85 committed:
Merge branch 'main' of https://github.com/bitwarden/clients into auth/beeep/zip-export

djsmith85 committed:
Separate AAA visually using whitespace within tests

djsmith85 committed:
Remove unused error var

djsmith85 committed:
Write test that verifies different http request failures when retrieving attachment data

djsmith85 committed:
Remove uneeded ignore lint rule

djsmith85 committed:
Rewrite test to actually check that ciphers assigned to an org are filtered out

djsmith85 committed:
Introduce ExportedVault return type (#13842)

* Define ExportedVault type unioned by 2 new types that describe a plain-text export vs a blob-based zip-export

* Extend static getFileName to handle formats and add unit-tests

* Introduce new export return type throughout the vault export module

- Update abstractions
- Update return types within implementations
- Update callers/consumers to handle the new return value
- Fix all unit tests

* Add support for new export return type and fix download of blobs via CLI

* Add documentation to public methods

---------

Co-authored-by: Daniel James Smith <djsmith85@users.noreply.github.com>
